File: apps/flutter/lib/domain/usecases.dart
/// Framework-free use cases: the business logic, unit-testable without Flutter.
library;

import '../core/core.dart';
import 'models.dart';
import 'repositories.dart';

/// Allowed reservation quantity, per the contract.
const reservationQuantityRange = (min: 1, max: 8);

/// Creates a hold with input hardening and idempotency.
///
/// The quantity is validated before anything is sent, so a bad value becomes a typed
/// [Validation] rather than a wasted round trip. The [idempotencyKey] is caller-owned and
/// must be stable across retries of the same intent, so a double tap makes one reservation.
class CreateReservationUseCase {
  final ReservationRepository _reservations;
  const CreateReservationUseCase(this._reservations);

  Future<Outcome<Reservation>> call(String sectorId, int quantity, String idempotencyKey) {
    if (quantity < reservationQuantityRange.min || quantity > reservationQuantityRange.max) {
      return Future.value(Failure(Validation(cause: 'quantity $quantity out of range')));
    }
    return _reservations.create(sectorId, quantity, idempotencyKey);
  }
}

/// Creates an order, mapping an unresolved create (timeout or network drop) to
/// [PaymentUnknown] rather than a false failure. Reporting failure here is how you
/// double-charge someone.
class CreateOrderUseCase {
  final OrderRepository _orders;
  const CreateOrderUseCase(this._orders);

  Future<Outcome<Order>> call(String reservationId, String idempotencyKey) async {
    final result = await _orders.create(reservationId, idempotencyKey);
    if (result is Failure<Order>) {
      final e = result.error;
      if (e is TimeoutError || e is NetworkUnavailable) {
        return Failure(PaymentUnknown(
          requestId: e.requestId,
          cause: 'order create unresolved (${e.code}); reconcile by retrying with the same key or polling',
        ));
      }
    }
    return result;
  }
}

/// The decision for one order-status poll.
sealed class Reconciliation {
  const Reconciliation();
}

class Resolved extends Reconciliation {
  final Order order;
  const Resolved(this.order);
}

class Continue extends Reconciliation {
  const Continue();
}

class Abort extends Reconciliation {
  final AppError error;
  const Abort(this.error);
}

/// Turns one order-status poll into a decision, with no timers. A settled order resolves; a
/// still-pending order or a transient failure continues; a non-transient error aborts. Kept
/// pure so the "when do we stop polling" logic is exhaustively testable.
Reconciliation reconcileOrderPoll(Outcome<Order> pollResult) {
  switch (pollResult) {
    case Success<Order>(value: final order):
      return order.isSettled ? Resolved(order) : const Continue();
    case Failure<Order>(error: final error):
      if (error is PaymentUnknown || error is TimeoutError || error is NetworkUnavailable || error is RateLimited) {
        return const Continue();
      }
      return Abort(error);
  }
}

File: apps/react-native/src/domain/usecases.ts
// Framework-free use cases: the business logic, unit-testable without React.

import { AppError, appError, fail, Outcome } from '../core/core';
import { isSettled, Order, Reservation } from './models';
import { OrderRepository, ReservationRepository } from './repositories';

export const RESERVATION_QUANTITY_RANGE = { min: 1, max: 8 } as const;

/**
 * Creates a hold with input hardening and idempotency. The quantity is validated before
 * anything is sent; the `idempotencyKey` is caller-owned and stable across retries of the
 * intent, so a double tap makes one reservation.
 */
export function createReservationUseCase(reservations: ReservationRepository) {
  return async (sectorId: string, quantity: number, idempotencyKey: string): Promise<Outcome<Reservation>> => {
    if (quantity < RESERVATION_QUANTITY_RANGE.min || quantity > RESERVATION_QUANTITY_RANGE.max) {
      return fail(appError('Validation', { cause: `quantity ${quantity} out of range` }));
    }
    return reservations.create(sectorId, quantity, idempotencyKey);
  };
}

/**
 * Creates an order, mapping an unresolved create (timeout or network drop) to
 * PaymentUnknown rather than a false failure. Reporting failure here is how you
 * double-charge someone.
 */
export function createOrderUseCase(orders: OrderRepository) {
  return async (reservationId: string, idempotencyKey: string): Promise<Outcome<Order>> => {
    const result = await orders.create(reservationId, idempotencyKey);
    if (!result.ok && (result.error.code === 'Timeout' || result.error.code === 'NetworkUnavailable')) {
      return fail(
        appError('PaymentUnknown', {
          requestId: result.error.requestId,
          cause: `order create unresolved (${result.error.code}); reconcile by retrying with the same key or polling`,
        }),
      );
    }
    return result;
  };
}

/** The decision for one order-status poll. */
export type Reconciliation =
  | { kind: 'resolved'; order: Order }
  | { kind: 'continue' }
  | { kind: 'abort'; error: AppError };

/**
 * Turns one order-status poll into a decision, with no timers. A settled order resolves; a
 * still-pending order or a transient failure continues; a non-transient error aborts. Pure,
 * so the "when do we stop polling" logic is exhaustively testable.
 */
export function reconcileOrderPoll(pollResult: Outcome<Order>): Reconciliation {
  if (pollResult.ok) {
    return isSettled(pollResult.value) ? { kind: 'resolved', order: pollResult.value } : { kind: 'continue' };
  }
  const code = pollResult.error.code;
  if (code === 'PaymentUnknown' || code === 'Timeout' || code === 'NetworkUnavailable' || code === 'RateLimited') {
    return { kind: 'continue' };
  }
  return { kind: 'abort', error: pollResult.error };
}
